### DISCUSSION TITLE
Challenge system rework

### DISCUSSION THREAD

**Discussion by RDIL** [thepeacockproject/Peacock#45] [Ideas]
So, right now, it's clear that challenges are just a mess.  They have multiple different locations that critical portions are stored, and as a result, they barely respect the proper rules that challenges should follow, as defined by IOI.  It's clear this system needs to be re-worked in a way that doesn't have the same flaws as the current one.  Hoping to use this discussion as a way to provide a forum for talking about the new design.

Some problems the current one faces:
- Data is duplicated across both memory and between different storage mediums
- Scopes ignored
- Different formats get quickly confusing

**Comment by moonysolari** [comment]
These three requests ought to have exactly the same data in the listed fields in their responses:
- `data.ChallengeData.Children` of `/profiles/page/DebriefingChallenges`,
- `data.ChallengeData.Children` of `/profiles/page/Planning`,
- `Data.Children` of `/authentication/api/userchannel/HubPagesService/GetChallengeTreeFor`.

So it is natural that the implementation of all three should call the same function to get this data. However, it is currently not the case:
- `/profiles/page/DebriefingChallenges` and `/authentication/api/userchannel/HubPagesService/GetChallengeTreeFor` use the `getChallengeTreeForContract` function.
- `/profiles/page/Planning` uses the `getChallengePlanningDataForContract` function, which calls the `reBatchIntoSwitchedData` function.

I took a look and there is quite a lot of code duplication between the two implementations. Refactoring this will likely fix #66 because the problem did not occur for debriefing challenges, in whose implementation I used the `getChallengeTreeForContract` function.
